HACKING.adoc: using dune to get merlin's support

trefis opened the pull request:
Ping @gasche.

If you want a travis check that this build keeps working, you'd need to tell travis to install the latest official release of the compiler, and dune, and then run `./configure && dune build @libs`.
But I don't know anything about travis, and not in the mood to learn about it right now. Any help would be appreciated.

In the mean time, hopefully the instructions are clear enough and work for you.

trefis committed:
HACKING.adoc: using dune to get merlin's support

gasche commented on HACKING.adoc:
```diff
@@ -302,6 +302,27 @@ If `boot/ocamlc` changes (e.g. because you ran `make bootstrap`), then
+
+Therefore, to use merlin on the compiler, you want to build the compiler with an
+older version of itself. The easiest way to do this is to install the latest
+released version of the compiler, install dune, and run:
```
If I understand correctly, you are papering over the fact that this only works when the compiler is buildable with the last released version (that is, when no bootstrap was in fact required since the last release). This will typically work well early in development cycles, and not at all late in development cycles. I think you should emphasize this, because as written the notes suggest that the method you propose should always work, while in fact we know it often does not. 

(Maybe one could mention the idea of developing compiler changes in the last release branch instead of trunk, as a way to avoid this problem and make it easy to test one's changes with the opam repository.)

trefis replied:
You do understand correctly, and I will add a note about that point, but I think you're painting a darker picture than necessary.
We do regularly need to bootstrap because of the way our build is setup: part of the distribution is built with the bootstrapped compiler, part with the compiler that was just built, and we're expecting the build artifacts of both compilers to be compatible. Which they regularly aren't.

But that does not mean that the distribution can't be fully built with an older compiler; for instance: `dune build @libs` on this branch from a 4.09 switch does work fine.

That being said: things do break sometimes, new features do end up being used in the compiler; and I will add some text about that.

gasche replied:
Ah, I hadn't realized the "compatible artifacts" thing (despite being the thing that is problematic with the no-flat-float-array kerkuffle right under my nose), thanks.

gasche commented on HACKING.adoc:
```diff
@@ -302,6 +302,27 @@ If `boot/ocamlc` changes (e.g. because you ran `make bootstrap`), then
+
+N.B. for this to work properly, you need to have run the configure script, and
+the workspace needs to be clean (otherwise dune will complain that build
+artefacts are present among the sources).
```
Correct me if I'm wrong, but the `.merlin` files will lookup the build artefacts generated by Dune, and not at all those generated by Make. This means that if (I am developing with Make and) I change a module API and I want Merlin to see the change, I need to run `make clean` first, and then do the dune thing. This should be clarified/documented, because otherwise users (including developers) are going to get confused.

Is there a way to ask Dune to ignore build artefacts, so that we don't have to make clean (and lose make incremental builds) to get Merlin to notice an API change?

trefis replied:
> Correct me if I'm wrong, but the .merlin files will lookup the build artefacts generated by Dune, and not at all those generated by Make. This means that if (I am developing with Make and) I change a module API and I want Merlin to see the change, I need to run make clean first, and then do the dune thing.

Correct.

> This should be clarified/documented, because otherwise users (including developers) are going to get confused.

I will try.

> Is there a way to ask Dune to ignore build artefacts, so that we don't have to make clean (and lose make incremental builds) to get Merlin to notice an API change?

Not to my knowledge. And I think it might not be an easy one to add. Let's ping @diml, see what he says.

gasche reviewed (commented):
Unless I misunderstood something, this approach will break once we make a compiler-distribution change which requires a bootstrap. If we want it in the CI, it should thus only be optional; I'm a bit unsure about having a CI test that is broken during a large part of the development cycle, and whose failure is completely unrelated to user's PRs.

trefis commented:
> If we want it in the CI, it should thus only be optional; I'm a bit unsure about having a CI test that is broken during a large part of the development cycle, and whose failure is completely unrelated to user's PRs.

I wholeheartedly agree with the "optional" part, not so much with the rest. I don't think the test is going to be broken for a large part of the development cycle.
All the issues I had with the dune build over the past 10 months were due to people adding files to the distribution and not updating the `dune` file. The optional CI would catch this issue.

Also, we could decide to disable that particular check until the next release as soon as there is a real breakage.

trefis committed:
HACKING.adoc: more explanations about the workflow

trefis committed:
HACKING.adoc: parting words

trefis commented:
(Doc updated)

gasche commented:
> Also, we could decide to disable that particular check until the next release as soon as there is a real breakage.

Yes, that sounds like a sensible approach.

trefis commented:
Ping @gasche.
If you don't want to look into setting anything up with travis, perhaps we should just merge this? (Assuming the text is now to your liking)

gasche commented on HACKING.adoc:
```diff
@@ -302,6 +302,39 @@ If `boot/ocamlc` changes (e.g. because you ran `make bootstrap`), then
+Finally, there will be times were the compiler simply cannot be built with an
+older version of itself. In which case, you'll either have to live without
+merlin, or first develop your patch on an older branch of the compiler before
+rebasing it (this can, however, introduce a substantial amount of extra work).
```
This can introduce a substantial amount of extra work (when later rebasing to the current development version). But it also makes it easier to test the impact of your work on third-party code, by installing a local opam switch: opam packages tend to be compatible with release versions of the compiler, whereas many of them are incompatible with the current development version.

gasche replied:
I pushed a version that includes my suggestion.

gasche commented:
@trefis I just pushed an attempt at a Travis check for this. I never played at depending on apt packages, and I am skeptical that the `ocaml` and `dune` versions packaged there will be good enough for this test, but I thought it would be worth a try.

You probably want to review the user-directed message I wrote in the travis test, which contributors will go read if there is a failure on their PR.

gasche commented:
@trefis the CI is now failing on a weird error that may be for you to fix. (I have no idea what's going on but at least Dune is running, so it may be a build-script-side issue.)

gasche commented:
I'm out of microbenchmarks so I fixed the dune build (in trunk) in 9eb9d047512c99aef5a749ad48bf2c3cda41229c.

gasche committed:
HACKING.adoc merlin: include the in-words advice in the code snippet as well

gasche commented:
This should be all working fine now.

trefis commented:
Thanks @gasche!
However after seeing #9519 I wonder if we should not be using github actions for the dune check. @dra27: any opinion on the matter?

trefis commented:
Ah but it wasn't merged, nevermind :)

trefis committed:
typo and wording

shindere commented:
Perhaps the travis part could be removed from this PR, the `HACKING.adoc`
related commits could be squashed and the PR could be merged that way?

gasche commented:
I agree that the HACKING.doc documentation has value and should be merged. That said, if I understand correctly the dune build does not currently work (for non-trivial reasons), so we are not in a hurry to encourage users to try it.

shindere commented:
Gabriel Scherer (2020/05/29 02:09 -0700):
> I agree that the HACKING.doc documentation has value and should be
> merged. That said, if I understand correctly the dune build does not
> currently work (for non-trivial reasons), so we are not in a hurry to
> encourage users to try it.

Indeed. There may still be value in taking out the Travis bit, though.

>

gadmm commented:
Thanks for your work on this.

As reported at #9662, I have trouble making it work, even after following these instructions. After installing a local opam switch using the latest official release, I managed to reduce the issue to 1 error and 3 warnings. The error is:
```
Unknown builtin primitive "%loc_FUNCTION"
```
so I guess 4.10 is still not recent enough.

- Are such errors expected and will Merlin still work partially?
- Is it possible to make it work with another environment? If so is there a general advice on which OCaml compiler to choose?

If the answer is yes to either question, I think this could make a good addition to the documentation. Currently, I do not know if I should try harder or if we are in the situation where _“the compiler simply cannot be built with an older version of itself”_ as you write.

gasche commented:
Yes, I believe that we are currently in this situation: a primitive that was introduced (by @nojb) in the current version of the compiler is used in the standard library, so we cannot build the standard library. Ideally we would:

1. document this failure mode better in the HACKING.adoc explanation, so that indeed you would know for sure that we are in this scenario
2. make it easy to build the compiler libraries with dune (the part we actually want Merlin for, most of the time) even when the standard library does not build for this reason

(Of course it is also possible that the compiler libraries themselves would be dune-broken by the use of a not-yet-released primitive; this is a best-effort thing.)

trefis commented:
I rebased now that the build was fixed, and removed the travis check.
The PR now only touches HACKING.adoc, @gasche do you agree this can now be merged?

Additionally, we should probably enable the check on inria's CI once again, but I'm not sure how to do that: I see the script in `tools/ci/inria` but what do I need to do for it to be run?

gasche commented:
Yes, I believe this can be merged. Regarding the CI, I think we should ask @xavierleroy or @shindere to take care of it. (My understanding is that they prefer to avoid other people touching the CI configuration in `trunk`.)

This being said, I think that the discussion with @gadmm above shows that we could/should be a bit more precise about the situations that are known to break the Dune build -- typically, a new primitive is introduced and used in the stdlib. This could be done by you as part of this PR.

I also think that it would be sensible to unplug the build of the standard library itself from `@libs` to avoid this issue better in the future. This is for the maintainers of the Dune build system in the compiler to change, and it doesn't need to be part of the present PR.

trefis commented:
> I also think that it would be sensible to unplug the build of the standard library itself from @libs to avoid this issue better in the future.

That's not possible, because the compiler needs to be build with the stdlib inside the `stdlib/` directory, because it depends on the definition of the format ast.
If you build the compiler with a preinstalled stdlib, and then someone adds a constructor to format (it happened a few versions ago, when I first wrote the dune files) then you can't build the compiler with dune anymore.

gasche commented:
That's fair, I guess? Some things just don't want to be nice.

I think that this would deserve a Changes entry, in part because it might be useful for other potential contributors to notice that it now exists.

shindere commented:
Thomas Refis (2020/06/24 08:27 -0700):
> Additionally, we should probably enable the check on inria's CI once
> again, but I'm not sure how to do that: I see the script in
> `tools/ci/inria` but what do I need to do for it to be run?

The job needs to be enabled from within Jenkins' web interface.

I can do it whenever you want, just ask.

trefis committed:
more explanations

trefis committed:
Changes

trefis commented:
@gasche: doc lightly updated to describe the failure that was experienced recently (we can expand as we encounter more / different failures), changes entry added.

shindere commented:
Thomas Refis (2020/06/25 05:11 -0700):
> Merged #9468 into trunk.

Dune-build Jenkins job re-enabled.

trefis commented:
Thank you @shindere !

shindere commented:
Thomas Refis (2020/06/25 06:49 -0700):
> Thank you @shindere !

My pleasure, you are very welcome.
